src/gdb_packet.c

```diff
@@ -31,119 +31,220 @@
+		// const char rx_char = state != PACKET_GDB_COMPLETE ? gdb_if_getchar() : '\0';
+		const char rx_char = gdb_if_getchar();
+		if (rx_char == '\x04') {
+			/* EOT (end of transmission) - connection was closed */
+			packet[0] = '\x04';
+			packet[1U] = 0;
+			return 1U;
+		}
+
+		switch (state) {
+		case PACKET_IDLE:
+			if (rx_char == GDB_PACKET_START) {
+				/* Start of GDB packet */
+				state = PACKET_GDB_CAPTURE;
 				offset = 0;
-				bool getting_remote_packet = true;
-				while (getting_remote_packet) {
-					/* Smells like bad code */
-					const char c = gdb_if_getchar();
-					switch (c) {
-					case REMOTE_SOM: /* Oh dear, packet restarts */
-						offset = 0;
-						break;
-
-					case REMOTE_EOM: /* Complete packet for processing */
-						packet[offset] = 0;
-						remote_packet_process(offset, packet);
-						getting_remote_packet = false;
-						break;
-
-					case '$': /* A 'real' gdb packet, best stop squatting now */
-						packet[0] = '$';
-						getting_remote_packet = false;
-						break;
-
-					default:
-						if (offset < size)
-							packet[offset++] = c;
-						else
-							/* Who knows what is going on...return to normality */
-							getting_remote_packet = false;
-						break;
-					}
-				}
-				/*
-				 * Reset the packet buffer start character to zero, because function
-				 * 'remote_packet_process()' above overwrites this buffer, and
-				 * an arbitrary character may have been placed there. If this is a '$'
-				 * character, this will cause this loop to be terminated, which is wrong.
+				checksum = 0;
+			}
+#if PC_HOSTED == 0
+			else if (rx_char == REMOTE_SOM) {
+				/* Start of BMP remote packet */
+				/* 
+				 * Let consume_remote_packet handle this
+				 * returns PACKET_IDLE or PACKET_GDB_CAPTURE if it detects the start of a GDB packet
 				 */
-				packet[0] = '\0';
+				state = consume_remote_packet(packet, size);
+				/* No need to handle checksum or offset, everything happened in PACKET_IDLE state so nothing was modified */
 			}
 #endif
-		} while (packet[0] != '$');
-
-		offset = 0;
-		csum = 0;
-		char c = '\0';
-		/* Capture packet data into buffer */
-		while (c != '#') {
-			c = gdb_if_getchar();
-			if (c == '#')
-				break;
-			/* If we run out of buffer space, exit early */
-			if (offset == size)
-				break;
+			break;
 
-			if (c == '$') { /* Restart capture */
+		case PACKET_GDB_CAPTURE:
+			if (rx_char == GDB_PACKET_START) {
+				/* Restart GDB packet capture */
 				offset = 0;
-				csum = 0;
-				continue;
+				checksum = 0;
+				break;
 			}
-			if (c == '}') { /* Escaped char */
-				c = gdb_if_getchar();
-				csum += c + '}';
-				packet[offset++] = (char)((uint8_t)c ^ 0x20U);
-				continue;
+			if (rx_char == GDB_PACKET_END) {
+				/* End of GDB packet */
+				// if (noackmode)
+				// 	/* knowlingly ignore the checksum that follows in no ack mode (as per spec) */
+				// 	state = PACKET_GDB_COMPLETE;
+				// else
+				/* Move to checksum capture */
```

Comment:
This should probably be changed to a block comment and the commented out dead code needs dropping

Reply:
This is part of a temporary commit to get things works, this is why I noted "theres still an issue" this wont be part of the final submission